back/engines/commercial/analysis/app/lib/analysis/auto_tagging_method/nlp_topic.rb

```diff
@@ -50,5 +55,18 @@ def parse_topic_modeling_response(response)
     def max_topics(inputs_count)
       [inputs_count, (Math.log(inputs_count, 5) * 6).ceil].min
     end
+
+    def fit_inputs_in_context_window(inputs, project_title)
+      prompt = inputs_prompt(inputs, project_title)
+      tokens_for_response = TOKENS_PER_TOPIC * max_topics(inputs.size)
+      tokens = LLM::AzureOpenAI.token_count(prompt) + tokens_for_response
+      exceeded_tokens = tokens - gpt4.context_window
+      return inputs if exceeded_tokens < 0
+
+      tokens_per_input = (LLM::AzureOpenAI.token_count(input_to_text.format_all(inputs)) / inputs.size.to_f).ceil
+      inputs_excess = (exceeded_tokens / tokens_per_input.to_f).ceil
+      inputs_excess = [inputs_excess, 1].max # Avoid infinite loop
+      fit_inputs_in_context_window(inputs.shuffle.drop(inputs_excess), project_title)
+    end
```

Comment:
It's true that it could be optimized, but I'm not sure it's worth spending more effort on it:
- Getting a response back takes a lot more time, making the performance gain insignificant
- It happens in an exceptional case. Ideally `fit_inputs_in_context_window` is only called once or twice, but we allow more recursive calls to have a stronger guarantee of remaining under the limit.
- I wanted to avoid coupling with the implementation of inputs_to_text (for example, making assumptions on how many tokens are used in the separators between the inputs).

Reply:
I hear your arguments, but I don't fully agree. I think the code is still too convoluted for what it does.